go/libraries/doltcore/sqle/dolt_diff_table_function.go

```diff
@@ -0,0 +1,458 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return dtf, nil
+}
+
+// Children implements the sql.Node interface
+func (dtf *DiffTableFunction) Children() []sql.Node {
+	return []sql.Node{}
+}
+
+// RowIter implements the sql.Node interface
+func (dtf *DiffTableFunction) RowIter(ctx *sql.Context, _ sql.Row) (sql.RowIter, error) {
+	err := dtf.evaluateArguments()
+	if err != nil {
+		return nil, err
+	}
+
+	if dtf.joiner == nil {
+		panic("schema and joiner haven't been initialized")
+	}
+
+	sqledb, ok := dtf.database.(Database)
+	if !ok {
+		panic("unable to get dolt database")
+	}
+	ddb := sqledb.GetDoltDB()
+
+	toRoot, toHash, toDate, err := dtf.loadDetailsForRef(ctx, dtf.toCommitVal, ddb)
+	if err != nil {
+		return nil, err
+	}
+
+	toTable, _, err := toRoot.GetTable(ctx, dtf.tableName)
+	if err != nil {
+		return nil, err
+	}
+
+	fromRoot, fromHash, fromDate, err := dtf.loadDetailsForRef(ctx, dtf.fromCommitVal, ddb)
+	if err != nil {
+		return nil, err
+	}
+
+	fromTable, _, err := fromRoot.GetTable(ctx, dtf.tableName)
+	if err != nil {
+		return nil, err
+	}
+
+	dp := dtables.NewDiffPartition(toTable, fromTable, toHash, fromHash, toDate, fromDate, &dtf.toSch, &dtf.fromSch)
+
+	return NewDiffTableFunctionRowIterForSinglePartition(*dp, ddb, dtf.joiner), nil
+}
+
+// loadDetailsForRef loads the root, hash, and timestamp for the specified ref value
+func (dtf *DiffTableFunction) loadDetailsForRef(ctx *sql.Context, ref interface{}, ddb *doltdb.DoltDB) (*doltdb.RootValue, string, *types.Timestamp, error) {
+	hashStr, ok := ref.(string)
+	if !ok {
+		return nil, "", nil, fmt.Errorf("received '%v' when expecting commit hash string", ref)
+	}
+
+	var root *doltdb.RootValue
+	var commitTime *types.Timestamp
+	cs, err := doltdb.NewCommitSpec(hashStr)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	cm, err := ddb.Resolve(ctx, cs, nil)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	root, err = cm.GetRootValue()
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	meta, err := cm.GetCommitMeta()
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	t := meta.Time()
+	commitTime = (*types.Timestamp)(&t)
+
+	return root, hashStr, commitTime, nil
+}
+
+// WithChildren implements the sql.Node interface
+func (dtf *DiffTableFunction) WithChildren(node ...sql.Node) (sql.Node, error) {
+	if len(node) != 0 {
+		panic("unexpected children")
+	}
+	return dtf, nil
+}
+
+// CheckPrivileges implements the sql.Node interface
+func (dtf *DiffTableFunction) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	// TODO: Implement priv check on underlying table
+	return true
+}
+
+// evaluateArguments evaluates the argument expressions to turn them into values this DiffTableFunction
+// can use. Note that this method only evals the expressions, and doesn't validate the values.
+func (dtf *DiffTableFunction) evaluateArguments() error {
+	if !dtf.Resolved() {
+		return nil
+	}
+
+	if !sql.IsText(dtf.tableNameExpr.Type()) {
```

Comment:
I think this check gives a slightly better error messages in some cases (e.g. passing a number for a commit gives a better error message with this check, otherwise it gives a confusing error message from "rootAsOf" about int8 not being valid for as of). I don't feel super strongly about it, though, so if you think it's really not worth it, I can take these checks out and just let arguments error out later.